utils/vcpkg_bootstrap.cmake

```diff
@@ -0,0 +1,106 @@
+    set(_baseline_sha "${_vcpkg_current_head_sha}")
+
+    file(READ "${CMAKE_SOURCE_DIR}/vcpkg-configuration.json" _vcpkg_configuration_json_contents)
+    string(JSON _default_registry_kind ERROR_VARIABLE _get_baseline_err GET "${_vcpkg_configuration_json_contents}" "default-registry" "kind")
+    if(_default_registry_kind STREQUAL "builtin")
+        string(JSON _maybe_baseline_sha ERROR_VARIABLE _get_baseline_err GET "${_vcpkg_configuration_json_contents}" "default-registry" "baseline")
+    elseif(_default_registry_kind STREQUAL "git")
+        string(JSON _git_registry_repo_url ERROR_VARIABLE _get_baseline_err GET "${_vcpkg_configuration_json_contents}" "default-registry" "repository")
+        if(_git_registry_repo_url STREQUAL "${X_VCPKG_GIT_REPOSITORY_URL}")
+            string(JSON _maybe_baseline_sha ERROR_VARIABLE _get_baseline_err GET "${_vcpkg_configuration_json_contents}" "default-registry" "baseline")
+        endif()
+    endif()
+    if(NOT _maybe_baseline_sha)
+        file(READ "${CMAKE_SOURCE_DIR}/vcpkg.json" _vcpkg_json_contents)
+        string(JSON _maybe_baseline_sha ERROR_VARIABLE _get_baseline_err GET "${_vcpkg_json_contents}" "builtin-baseline")
+    endif()
+
+    if(_maybe_baseline_sha)
+        set(_baseline_sha "${_maybe_baseline_sha}")
+    endif()
+
+    set(_should_update_vcpkg FALSE)
+    if(NOT "${_vcpkg_current_head_sha}" STREQUAL "${_baseline_sha}")
+        set(_should_update_vcpkg TRUE)
+    endif()
+    foreach(_patch IN LISTS VCPKG_PATCHES)
+      get_filename_component(_patch_name "${_patch}" NAME)
+      if(NOT EXISTS "${X_VCPKG_CLONE_DIR}/${_patch_name}.applied")
+        set(_should_update_vcpkg TRUE)
+        break()
+      endif()
+    endforeach()
+
+    if(_should_update_vcpkg)
+        message(STATUS "Fetching changes from vcpkg upstream")
+        execute_process(
+            COMMAND "${GIT_EXECUTABLE}" fetch --quiet --prune
+            WORKING_DIRECTORY "${X_VCPKG_CLONE_DIR}"
+            # No error checking here to allow offline usage
+        )
+        message(STATUS "Fetching changes from vcpkg upstream - done")
+
+        message(STATUS "Switching vcpkg HEAD to ${_baseline_sha}")
+        execute_process(
+            COMMAND "${GIT_EXECUTABLE}" reset --quiet --hard "${_baseline_sha}"
+            COMMAND "${GIT_EXECUTABLE}" clean -qfd
+            WORKING_DIRECTORY "${X_VCPKG_CLONE_DIR}"
+            COMMAND_ERROR_IS_FATAL ANY
+        )
+        message(STATUS "Switching vcpkg HEAD to ${_baseline_sha} - done")
+
+        # Apply patches
+        foreach(_patch IN LISTS VCPKG_PATCHES)
+            message(STATUS "Applying patch ${_patch}")
+            execute_process(
+                COMMAND "${GIT_EXECUTABLE}" apply --whitespace nowarn "${_patch}"
+                COMMAND "${CMAKE_COMMAND}" -E touch "${X_VCPKG_CLONE_DIR}/${_patch_name}.applied"
+                WORKING_DIRECTORY "${X_VCPKG_CLONE_DIR}"
+                COMMAND_ERROR_IS_FATAL ANY
+            )
+        endforeach()
```

Comment:
A lot of lines here could probably be safely removed for better maintainability. We don't have an vcpkg patches, so all lines concerning them are moot. Likewise for any registry type other than `builtin`.